larpandoracontent/LArHelpers/LArHierarchyHelper.h

```diff
@@ -0,0 +1,629 @@
+            int m_pdg;
+        };
+
+        /**
+         *  @brief  Default constructor
+         */
+        MCHierarchy() = default;
+
+        /**
+         *  @brief  Construct a new MCHierarchy object using specified reconstructability criteria
+         *
+         *  @param  recoCriteria The reconstructability criteria to be applied
+         */
+        MCHierarchy(const ReconstructabilityCriteria &recoCriteria);
+
+        /**
+         *  @brief Destructor
+         */
+        virtual ~MCHierarchy();
+
+        /**
+         *  @brief  Creates an MC hierarchy representation. Without folding this will be a mirror image of the standard MCParticle
+         *          relationships. However, with folding options selected the hierarchy structure will group together MC particles into
+         *          nodes based on the folding requirements.
+         *
+         *          If only folding back to primaries, the hierarchy will be relatively flat, with a top-level neutrino or test beam
+         *          particle, if appropriate, and then a set of leaf nodes, one for each primary particles also containing the MC particles
+         *          (and corresponding hits) from daughter particles.
+         *
+         *          If only folding back to leading shower particles, the hierarchy will largely mirror the standard MCParticle hierarchy,
+         *          but, when a shower particle is reached (for this purpose an electron or photon), this particle and all daughter
+         *          particles will be represented by a single leaf node.
+         *
+         *          If folding back to both primary and leading shower particles the hierarchy will again be rather flat, but in this case,
+         *          if a primary track-like particle (i.e. not an electron or photon) has a downstream shower particle then all downstream
+         *          particles above the shower-like particle will be folded into the primary node, but a new, daughter leaf node will be
+         *          created for the shower-like particle and all of its daughters, and a parent-child relationship will be formed between
+         *          the primary node and shower node.
+         *
+         *  @param  mcParticleList The list of MC particles with which to fill the hierarchy
+         *  @param  caloHitList The list of hits with which to fill the hierarchy
+         *  @param  foldToPrimaries Whether or not to fold daughter particles back to their primary particle
+         *  @param  foldToLeadingShowers Whether or not to fold daughter particles back to their leading shower particle
+         */
+        void FillHierarchy(const pandora::MCParticleList &mcParticleList, const pandora::CaloHitList &caloHitList,
+            const bool foldToPrimaries, const bool foldToLeadingShowers);
+
+        // Note - likely want to change this to retrieve the neutrino, the test beam particle, all primaries, all cosmics etc
```

Comment:
I've now removed this comment. I was imagining a slightly different implementation when I wrote this, which no longer applies. That said, it might nonetheless be useful to provide additional functions to extract only cosmics, or only test beam particles - I'm undecided on this, as it's actually not needed from the perspective of matching, so I've omitted it for now, but could add this in the future if people think it's particularly useful.